File: java-client/src/main/java/com/couchbase/client/java/query/ReactiveQueryResult.java
/*
 * Copyright (c) 2019 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.java.query;

import com.couchbase.client.core.annotation.Stability;
import com.couchbase.client.core.error.DecodingFailureException;
import com.couchbase.client.core.msg.query.QueryResponse;
import com.couchbase.client.java.codec.JsonSerializer;
import com.couchbase.client.java.codec.TypeRef;
import com.couchbase.client.java.json.JsonObject;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

/**
 * Reactive result that fetch parts of the N1ql Query responses reactively
 *
 * @since 3.0.0
 */
public abstract class ReactiveQueryResult {

	/**
	 * The default serializer to use.
	 */
	protected final JsonSerializer serializer;

	@Stability.Internal
	public ReactiveQueryResult(final JsonSerializer serializer) {
		this.serializer = serializer;
	}

	/**
	 * Get a {@link Flux} which publishes the rows that were fetched by the query which are then decoded to
	 * {@link JsonObject}
	 *
	 * @return {@link Flux}
	 */
	public Flux<JsonObject> rowsAsObject() {
		return rowsAs(JsonObject.class);
	}

	/**
	 * Get a {@link Flux} which publishes the rows that were fetched by the query which are then decoded to the
	 * requested entity class
	 *
	 * @param target target class for converting the query row
	 * @return {@link Flux}
   * @throws DecodingFailureException (async) if the decoding cannot be completed successfully
	 */
	public abstract <T> Flux<T> rowsAs(Class<T> target);

  /**
   * Get a {@link Flux} which publishes the rows that were fetched by the query which are then decoded to the
   * requested entity type
   *
   * @param target target type for converting the query row
   * @return {@link Flux}
   * @throws DecodingFailureException (async) if the decoding cannot be completed successfully
   */
	public abstract <T> Flux<T> rowsAs(TypeRef<T> target);

	/**
	 * Returns a {@link Mono} containing a {@link QueryMetaData},  giving access to the additional metadata associated with
	 * this query.
	 *
	 * Note that the metadata will only be available once all rows have been received, so it is recommended that you
	 * first handle the rows in your code, and then the metadata.  This will avoid buffering all the rows in-memory.
	 */
	public abstract Mono<QueryMetaData> metaData();
}

File: scala-client/src/main/scala-2/com/couchbase/client/scala/query/ReactiveQueryResult.scala
/*
 * Copyright (c) 2025 Couchbase, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.couchbase.client.scala.query

import com.couchbase.client.core.deps.io.netty.util.CharsetUtil
import com.couchbase.client.core.error.CouchbaseException
import com.couchbase.client.core.msg.query.QueryChunkRow
import com.couchbase.client.core.util.Golang
import com.couchbase.client.scala.codec.JsonDeserializer
import com.couchbase.client.scala.json.{JsonObject, JsonObjectSafe}
import com.couchbase.client.scala.util.{DurationConversions, RowTraversalUtil}
import reactor.core.scala.publisher.{SFlux, SMono}

import scala.concurrent.duration.Duration
import scala.util.{Failure, Success, Try}

/** The results of a N1QL query, as returned by the reactive API.
  *
  * @param metaData            any additional information related to the query
  */
case class ReactiveQueryResult(
    private[scala] val rows: SFlux[QueryChunkRow],
    metaData: SMono[QueryMetaData]
) {

  /** A Flux of any returned rows, streamed directly from the query service.  If the query service returns an error
    * while returning the rows, it will be raised on this.
    *
    * @tparam T any supported type; see [[https://docs.couchbase.com/scala-sdk/current/howtos/json.html these JSON docs]] for a full list
    */
  def rowsAs[T](implicit deserializer: JsonDeserializer[T]): SFlux[T] = {
    rows.map(row => {
      // The .get will raise an exception as .onError on the flux
      deserializer.deserialize(row.data()).get
    })
  }
}
